File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/configurations/common/before_launch/ConfigurationTypeMapper.java
package com.github.alexandrecarlton.idea.settings.applier.impl.configurations.common.before_launch;

import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.RunConfigurationType;
import com.intellij.execution.configurations.ConfigurationType;
import com.intellij.execution.remote.RemoteConfigurationType;
import java.util.function.Supplier;
import javax.inject.Inject;
import javax.inject.Named;

public class ConfigurationTypeMapper {

  private final Supplier<ConfigurationType> dockerConfigurationTypeSupplier;
  private final Supplier<ConfigurationType> shellScriptConfigurationTypeSupplier;
  private final Supplier<ConfigurationType> springBootConfigurationTypeSupplier;

  @Inject
  public ConfigurationTypeMapper(@Named("Docker") Supplier<ConfigurationType> dockerConfigurationTypeSupplier,
                                 @Named("Shell Script") Supplier<ConfigurationType> shellScriptConfigurationTypeSupplier,
                                 @Named("Spring Boot") Supplier<ConfigurationType> springBootConfigurationTypeSupplier) {
    this.dockerConfigurationTypeSupplier = dockerConfigurationTypeSupplier;
    this.shellScriptConfigurationTypeSupplier = shellScriptConfigurationTypeSupplier;
    this.springBootConfigurationTypeSupplier = springBootConfigurationTypeSupplier;
  }

  public ConfigurationType mapRunConfigurationType(RunConfigurationType runConfigurationType) {
    switch (runConfigurationType) {
      case DOCKER: return dockerConfigurationTypeSupplier.get();
      case REMOTE: return RemoteConfigurationType.getInstance();
      case SHELL_SCRIPT: return shellScriptConfigurationTypeSupplier.get();
      case SPRING_BOOT: return springBootConfigurationTypeSupplier.get();
      default:
        throw new IllegalArgumentException("Unhandled RunConfigurationType: " + runConfigurationType);
    }
  }
}

File: src/main/java/com/github/alexandrecarlton/idea/settings/applier/impl/configurations/common/before_launch/ConfigurationTypeMapper.kt
package com.github.alexandrecarlton.idea.settings.applier.impl.configurations.common.before_launch

import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.RunConfigurationType
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.RunConfigurationType.DOCKER
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.RunConfigurationType.REMOTE
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.RunConfigurationType.SHELL_SCRIPT
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.RunConfigurationType.SPRING_BOOT
import com.intellij.execution.configurations.ConfigurationType
import com.intellij.execution.remote.RemoteConfigurationType
import java.util.function.Supplier
import javax.inject.Inject
import javax.inject.Named

class ConfigurationTypeMapper @Inject
constructor(@param:Named("Docker") private val dockerConfigurationTypeSupplier: Supplier<ConfigurationType>,
            @param:Named("Shell Script") private val shellScriptConfigurationTypeSupplier: Supplier<ConfigurationType>,
            @param:Named("Spring Boot") private val springBootConfigurationTypeSupplier: Supplier<ConfigurationType>) {

    fun mapRunConfigurationType(runConfigurationType: RunConfigurationType) =
        when (runConfigurationType) {
            DOCKER -> dockerConfigurationTypeSupplier.get()
            REMOTE -> RemoteConfigurationType.getInstance()
            SHELL_SCRIPT -> shellScriptConfigurationTypeSupplier.get()
            SPRING_BOOT -> springBootConfigurationTypeSupplier.get()
        }
}
